interfaces/builtin/utils.go

```diff
@@ -125,3 +126,49 @@ func verifySlotPathAttribute(slotRef *interfaces.SlotRef, attrs interfaces.Attre
 	}
 	return cleanPath, nil
 }
+
+// aareExclusivePatterns takes a string and generates deny alternations. Eg,
+// aareExclusivePatterns("foo") returns:
+// []string{
+//   "[^f]*",
+//   "f[^o]*",
+//   "fo[^o]*",
+// }
+func aareExclusivePatterns(orig string) []string {
+	s := make([]string, len(orig))
+	prefix := ""
+	for i, letter := range orig {
+		if i > 0 {
+			prefix = orig[:i]
+		}
+		s[i] = fmt.Sprintf("%s[^%c]*", prefix, letter)
```

Comment:
since apparmor globbing is not quite regex, are there any characters here we might need to escape such as "^" itself or is this all handled in the expected way by apparmor's regex/globs ?

Reply:
@jdstrand is the authority here, but I suspect the problematic characters would be  `\`, `]`, `"`, and maybe newline.  This could probably be handled by checking whether `orig` matches a regexp searching for the problem characters.

Given that the input to this function is validated to not contain the problem characters, I don't think it is worth trying to implement escaping.

Reply:
> Given that the input to this function is validated to not contain the problem characters, I don't think it is worth trying to implement escaping.

That is precisely why I did not do this; we are using this with snapInstanceName only. I'll add a snap.ValidateInstanceName() and add some additional tests.

Reply:
Done